### ISSUE TITLE
[duckdb] sqlfluff fails to parse expressions within unpivots

### ISSUE THREAD

**Issue by paultiq** [sqlfluff/sqlfluff#7391]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Using first example from https://duckdb.org/docs/stable/sql/statements/unpivot#expressions-within-unpivot-statements: 

SQLFluff Version 4.0.0

### Expected Behaviour

Should parse properly.

### Observed Behaviour

```py
❯ uv run sqlfluff lint --dialect duckdb test1.sql
== [test.sql] FAIL
L:   1 | P:   1 | LT13 | Files must not begin with newlines or whitespace.
                       | [layout.start_of_file]
L:   2 | P:   1 |  PRS | Line 2, Position 1: Found unparsable section: "UNPIVOT\n
                       | (SELECT 42 AS col1, 'woot' A..."
WARNING: Parsing errors found and dialect is set to 'duckdb'. Have you configured your dialect correctly?
All Finished 📜 🎉!
```

```py
uv run sqlfluff lint --dialect duckdb test2.sql
== [test.sql] FAIL
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: "with q1
                       | as \n(\n    select 'A' x, * from r..."
WARNING: Parsing errors found and dialect is set to 'duckdb'. Have you configured your dialect correctly?
All Finished 📜 🎉!
```

### How to reproduce

test1.sql:
```sql
UNPIVOT
    (SELECT 42 AS col1, 'woot' AS col2)
    ON
        (col1 * 2)::VARCHAR,
        col2;
```

test2.sql:
```sql
with q1 as 
(
    select 'A' x, * from range(10) t(r)
)
unpivot q1 ON columns('r')::float4 AS '\1_\2'
```

### Dialect

Duckdb

### Version

4.0.0

### Configuration

None, in this case.

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
